[Merged by Bors] - feat(SymbolicDynamics): basic setup of Zd, full shift, cylinders, pat…

Sfgangloff opened the pull request:
This PR adds a **group-generic** foundation for symbolic dynamics over an arbitrary group `G`, together with convenient specializations for `ℤ` and `ℤ^d`.

Summary of additions:

- **Full shift and shift action**
  - `abbrev FullShift (A G) := G → A` (inherits product topology from the Π-type).
  - Right shift `shift g x` with convention `(shift g x) h = x (h * g)`.

- **Cylinders and topology**
  - `cylinder U x : Set (G → A)` for finite `U : Finset G`.
  - Cylinders are open under `[DiscreteTopology A]`; with a finite alphabet they are also closed.
  - Equality with dependent products:  
    `cylinder U x = Set.pi (↑U) (fun i => ({x i} : Set A))`, enabling use of the `Set.pi` API.

- **Patterns, occurrences, and subshifts**
  - `Pattern A G` with finite `support : Finset G` and `data : support → A`.
  - `Pattern.occursIn p x g` (occurrence at translate `g`) and the expected shift law.
  - `forbids F` and `Subshift A G` (closed, shift-invariant subsets).
  - `FixedSupport A G U` with an equivalence to `(U → A)` to obtain finiteness.

- **Language on finite shapes and counting**
  - `languageOn X U`, `languageCardOn X U`, and `patternCountOn Y U`.

- **Entropy along a shape sequence**
  - `limsupAtTop` (as an `sInf` of eventual upper bounds).
  - `entropyAlong X F hF := limsup (log (patternCountOn X (F n) + 1) / |F n|)`  
    for any nonempty finite shapes `F : ℕ → Finset G` (the `+ 1` avoids `log 0`).

- **Specializations**
  - `IntShapes`: segments `[-n,n]` on `Multiplicative ℤ`, with `entropy_Z`.
  - `ZdShapes`: boxes `[-n,n]^d` on `ℤ^d` (as functions `Fin d → ℤ`), with `entropy_Zd`.

Mathematical remarks:

- The API is **shape-parametric**: entropy is defined along user-provided finite shapes.
- On **amenable** groups, using a **Følner** sequence yields a canonical value (Ornstein–Weiss).  
  This PR does not assume amenability; the family of shapes is an explicit input.

Motivation:

Provide a clean, reusable base for symbolic dynamics on groups in mathlib.

Future work:

- Add a Følner predicate and prove shape-independence / limit existence on amenable groups.
- Expand the `ℤ`/`ℤ^d` toolkit (alternative shapes, mixing, factors).
- Develop 1D theory and, longer-term, multidimensional SFT results (e.g. along the lines of Hochman–Meyerovitch).

Commit:
feat(SymbolicDynamics): basic setup of Zd, full shift, cylinders, patterns, subshifts, entropy

sgouezel commented:
It's probably not a good idea to do this over `Fin d → ℤ`, as this will not apply to subshifts over `ℤ` (since `Fin 1 → ℤ` and `ℤ` are not the same thing). A solution here would be to abstract away, i.e., do symbolic dynamics over a group `G` (where you can instantiate `G` to be either `Fin d → ℤ` or `ℤ`). I don't expect the general theory to be more complicated than the one you develop (possibly simpler, in fact). You might need at some point to assume that `G` is finitely generated, but don't do it before you need it!

kckennylau commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,357 @@
+
+variable {d : ℕ}
+
+/-- The lattice `ℤ^d` as functions `Fin d → ℤ`. -/
+def Zd (d : ℕ) := Fin d → ℤ
```
(Update: this comment was made before I saw @sgouezel 's comment above. His comment supersedes mine.)

Sfgangloff replied:
Noted, thank you! I see your point. I originally introduced it as a def for readability, but I agree that an abbrev (or even just a notation) makes more sense here, since it’s only a type synonym. I will follow Sébastien’s suggestion from above and update accordingly.

Sfgangloff commented:
> It's probably not a good idea to do this over `Fin d → ℤ`, as this will not apply to subshifts over `ℤ` (since `Fin 1 → ℤ` and `ℤ` are not the same thing). A solution here would be to abstract away, i.e., do symbolic dynamics over a group `G` (where you can instantiate `G` to be either `Fin d → ℤ` or `ℤ`). I don't expect the general theory to be more complicated than the one you develop (possibly simpler, in fact). You might need at some point to assume that `G` is finitely generated, but don't do it before you need it!

Thank you for the advice ! I did not want to do it in full generality at first, but indeed d=1 is an important case. I'll try revamp this for groups in general.

Commit:
chore: update import all files

Commit:
symbolic dynamics: fix header/docstring order; cylinder lemmas; classical instances

eric-wieser commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,396 @@
+def FullShiftZd (A : Type*) (d : ℕ) := Zd d → A
+
+/-- Product topology on the full shift. -/
+instance : TopologicalSpace (FullShiftZd A d) := Pi.topologicalSpace
```
Because you used `reducible` above, you don't get to choose this, it's forced upon you automatically. I'd suggest just deleting this instance, and same for the one below

Commit:
refactor(symbolic-dynamics): group-generic subshifts, abbrev FullShift, cleanup cylinders and instances

- use abbrev FullShift := G → A
- remove redundant topology/inhabited instances
- relate cylinder to Set.pi
- reorganize lemmas to avoid unused variables
- separate ℤ^d-specific geometry/entropy

Commit:
chore: update import-all (lake exe mk_all)

Commit:
typo

Commit:
refactor(symbolic-dynamics): cylinder_is_open via isOpen_set_pi using cylinder_eq_set_pi (no logic change)

Commit:
chore(symbolic-dynamics): replace Pi/Discrete imports with Topology.Constructions

Commit:
refactor(symbolic-dynamics): prove  via

Commit:
changed full shift to G -> A and wrote the full shift as example of subshift

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+
+This file develops a minimal API for symbolic dynamics over an arbitrary group `G`.
+The ambient space is the full shift
+`FullShift A G := G → A` (an `abbrev`), hence it inherits the product topology
```
Could you drop all references to `FullShift` in the file?

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+  { y | ∀ i ∈ U, y i = x i }
+
+lemma cylinder_eq_set_pi (U : Finset (G)) (x : G → A) :
+  cylinder U x = Set.pi (↑U : Set (G)) (fun i => ({x i} : Set A)) := by
```
```suggestion
    cylinder U x = Set.pi (↑U : Set (G)) (fun i => ({x i} : Set A)) := by
```
Statements are indented by 4 spaces.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+
+section ShiftTopology  -- add only topology on A
+variable {A G : Type*} [Group G] [TopologicalSpace A]
+lemma shift_continuous (g : G) : Continuous (shift (A:=A) (G:=G) g) := by
```
```suggestion
lemma continuous_shift (g : G) : Continuous (shift (A := A) (G := G) g) := by
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+/-! ## Full shift and shift action -/
+
+
+/-- Right-translation shift: `(shift g x)(h) = x (h * g)`. -/
```
```suggestion
/-- Right-translation shift: `(shift g x) (h) = x (h * g)`. -/
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+
+
+/-- Right-translation shift: `(shift g x)(h) = x (h * g)`. -/
+def shift (g : G) (x : G → A) : G → A :=
```
Down the road, you will want to apply this mainly to additive groups, like Z or Z^d. Currently, your framework is only done for multiplicative groups. It is probably a good idea to do it simultaneously for multiplicative and additive groups. We have a device for this, called `to_additive`: do everything in the multiplicative setting, but tag definitions and results for `to_additive` and it will create automatically also the additive version.

Some care should be taken with names. For instance, you could use `mulShift` for the shift in a multiplicative context, and `addShift` in an additive context (if you tag a definition called `mulShift` with `to_additive`, it should pick automatically the name `addShift` for the additivized version)

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+
+namespace SymbolicDynamics
+
+variable {A : Type*} [Fintype A] [Fintype A] [DecidableEq A] [Inhabited A]
```
Only put the typeclasses you really need here. I don't see why you would need `Fintype` of `DecidableEq` or `Inhabited` from the start. If they are necessary at some point, only add them here.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+
+@[simp] lemma mem_cylinder {U : Finset G} {x y : G → A} :
+  y ∈ cylinder U x ↔ ∀ i ∈ U, y i = x i := Iff.rfl
+end CylindersDefs
```
```suggestion

end CylindersDefs
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+  ext y; simp [cylinder, Set.pi, Finset.mem_coe]
+
+@[simp] lemma mem_cylinder {U : Finset G} {x y : G → A} :
+  y ∈ cylinder U x ↔ ∀ i ∈ U, y i = x i := Iff.rfl
```
```suggestion
    y ∈ cylinder U x ↔ ∀ i ∈ U, y i = x i := Iff.rfl
```
Four spaces in theorem statements

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+
+section CylindersOpen
+variable {A G : Type*} [TopologicalSpace A] [DiscreteTopology A]
+/-- Cylinders are open (and, dually, closed) when `A` is discrete. -/
```
```suggestion
/-- Cylinders are open (and, dually, closed) when `A` is discrete. -/

```
Lemmas should have spaces before and after them.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+variable {A G : Type*} [TopologicalSpace A] [DiscreteTopology A]
+/-- Cylinders are open (and, dually, closed) when `A` is discrete. -/
+lemma cylinder_is_open (U : Finset G) (x : G → A) :
+  IsOpen (cylinder (A:=A) (G:=G) U x) := by
```
```suggestion
  IsOpen (cylinder (A := A) (G := G) U x) := by
```
Spaces around `:=`.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+/-- Cylinders are open (and, dually, closed) when `A` is discrete. -/
+lemma cylinder_is_open (U : Finset G) (x : G → A) :
+  IsOpen (cylinder (A:=A) (G:=G) U x) := by
+  classical
```
proofs should only be indented by 2 spaces

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,316 @@
+                     (t := fun i => ({x i} : Set A))) :=
+    isOpen_set_pi (U.finite_toSet) hopen
+  simpa [cylinder_eq_set_pi (A:=A) (G:=G) U x] using hpi
+end CylindersOpen
```
```suggestion

end CylindersOpen
```

sgouezel reviewed (commented):
Thanks for your PR! I have mostly made form comments in my review, to show you how your code should be formatted for mathlib. Could you try to use there rules also for the rest of the PR?

Also, reviewing long PRs like this one is quite complicated. Could you maybe split it, postponing the entropy part to another PR, to make sure that we can get your PR in quicker? Thanks!

Commit:
chore(Dynamics/SymbolicDynamics/Basic): drop  references

Commit:
style(Dynamics/SymbolicDynamics/Basic): fix indentation in lemmas

Commit:
style(Dynamics/SymbolicDynamics/Basic): add spaces around  in variable declarations

Commit:
style(Dynamics/SymbolicDynamics/Basic): apply remaining style changes

Commit:
style(Dynamics/SymbolicDynamics/Basic): removed entropy part

Commit:
refactor(Dynamics/SymbolicDynamics/Basic): prepare and add additive API via

Commit:
chore(Dynamics/SymbolicDynamics): re-add stub  to satisfy import

Commit:
removed entropy file

Commit:
chore(SymbolicDynamics): deglobalize new simp lemmas to avoid  churn

Commit:
chore(SymbolicDynamics): fix lints — docstrings for additive decls; silence unused args in fullShift/addFullShift

Commit:
chore(SymbolicDynamics): deglobalize simp on  to avoid  churn

sgouezel commented:
There are some linting errors, but certainly not related to your PR. Could you merge master to fix them?

Commit:
linters docstrings

Sfgangloff commented:
> There are some linting errors, but certainly not related to your PR. Could you merge master to fix them?

Done

Sfgangloff commented:
> My question is mainly for the map itself: if you have a subset `s`, the shift does not in general restrict to the subset, you need invariance. Also, for cylinders, viewing them inside the subtype means some more work (you need to take the preimage of the cylinder under the inclusion map). I am worried that this will make working with subshifts more awkward than it should be. If you have already worked out more of the theory and if you have an example to show me that everything works smoothly, that's great -- could you give me a pointer?

Thanks for clarifying! I see your point: if one defines cylinders and the shift only “inside a subshift,” then one gets nice ergonomics locally, but there is overhead whenever combining or comparing different subshifts. For example, suppose I want to prove that
\mathcal L_U(X \cup Y) = \mathcal L_U(X) \cup \mathcal L_U(Y).
Even if languages were defined relative to a subshift, in such a statement I would still need to move both sides back to the ambient pattern type in G \to A to take the union. The same issue arises with intersections, factors, products, etc. In that sense, having the ambient definitions as the “ground truth” is what makes these set-theoretic identities essentially tautological.

In the current design, all definitions (shift, cylinder, occurrence, language, …) live in the ambient full shift, and a subshift is just a closed, invariant subset. Then for a subshift Y I can view its language as 

languageOn Y.carrier U 

and results like the union lemma become immediate. When needed for ergonomics, I can certainly add thin wrappers (e.g. “language of a subshift on U”), but they would just unfold to the ambient versions.

So I think your concern is exactly the tradeoff: restricting everything to subshifts makes the local API slightly nicer but the global manipulations harder. Keeping things ambient keeps the global operations trivial, and I expect most interesting theory (SFTs, complexity, entropy, …) will want to combine subshifts.

I don’t yet have a long worked-out example, but even at this basic stage you can already see the pattern with unions. If you’d like, I can include some of these wrapper lemmas in the PR so that the ergonomics for subshifts are visible right away.

sgouezel commented:
I see your point. I think it's just a difference of background: from my point of view, I work a lot inside a fixed subshift (to study entropy, transfer operators, thermodynamic formalism, and so on), but I essentially never work in the space of subshifts (i.e., I'll never take the union of two subshifts).

Since you are the one doing the PR, we should go with your point of view. Could you please explain this design motivation in the file-level docstring? Also, it might be a good idea to namespace everything, to make sure that if we want to define cylinders with the inner point of view we don't get a clash.

Commit:
docstrings + ambiant namespace

Commit:
feat(SymbolicDynamics): support cancellative monoids, docstring refactor

* Generalized all group assumptions to
  instead of . This allows working with cancellative monoids,
  not only groups.
* Reworked  to avoid inverses, using right-cancellation
  instead. Added lemma  to characterize
  its behavior on translated supports.
* Fixed proofs depending on inverses () to use
  cancellation arguments.
* Expanded the file-level docstring: clarified assumptions (cancellative monoid instead of group)

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,511 @@
+
+variable {A : Type*}
+variable {G : Type*}
+variable [Group G]
```
Makes sense. I replaced Group by Monoid everywhere, some places require [IsRightCancelMul G] on top of [Monoid G] though.

Sfgangloff commented:
> Once you're happy, can you remove the awaiting-author tag? Otherwise, the PR does not appear in the queue, so I'm very likely to forget it!

@sgouezel I don’t seem to have permission to remove the awaiting-author tag (maybe because I’m marked as a new contributor ?). The PR is ready for another review though.

sgouezel commented:
You need to post a message with `-awaiting-author`.

Commit:
linters

Commit:
linters

Sfgangloff commented:
> You need to post a message with `-awaiting-author`.

-awaiting-author

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+For `Y : Subshift A G`, cylinders and occurrence sets *inside `Y`* are simply
+preimages of the ambient ones under the inclusion `Y → (G → A)`. For example:
+
+`{ y : Y | ∀ i ∈ U, (y : G → A) i = (x : G → A) i } = (Subtype.val) ⁻¹’ (cylinder U (x : G → A)).`
```
```suggestion
`{ y : Y | ∀ i ∈ U, (y : G → A) i = (x : G → A) i } = (Subtype.val) ⁻¹' (cylinder U (x : G → A)).`
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+
+namespace SymbolicDynamics
+
+namespace Ambiant
```
In English, I think it's Ambient, not Ambiant. But I think I'd just call the namespace FullShift, because the meaning of Ambient is essentially impossible to guess. And the inner one could be called Subshift

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+variable {G : Type*}
+variable [Monoid G]
+
+/-- Right-translation shift:
+* In multiplicative notation: `(mulShift g x) h = x (h * g)`.
+* In additive notation (e.g. for `ℤ^d`): `(addShift v x) u = x (u + v)` -/
```
You should have one docstring for the multiplicative notion, only mentioning multiplicative stuff. And another one for the additive notion, only mentioning additive stuff, and given in the `to_additive` call.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+
+attribute [inherit_doc SymbolicDynamics.Ambiant.mulShift] SymbolicDynamics.Ambiant.addShift
+
+end ShiftDefinition
+
+section ShiftAlgebra
```
If you have two consecutive sections that have exactly the same variables, it's better to merge them into one single section. I'd also merge them with the topological section, adding `[TopologicalSpace A]` when it becomes relevant.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+
+variable {A G : Type*}
+
+/-- Cylinder fixing `x` on the finite set `U`. -/
```
This docstring is not really good, as people who don't know what a cylinder is can not guess it from the docstring. Could you try to improve it?

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+variable {A G : Type*} [TopologicalSpace A] [DiscreteTopology A]
+
+/-- Cylinders are open (and, dually, closed) when `A` is discrete. -/
+lemma cylinder_is_open (U : Finset G) (x : G → A) :
```
A lemma name should follow the symbols used in the statement. So,
```suggestion
lemma isOpen_cylinder (U : Finset G) (x : G → A) :
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+
+end CylindersDefs
+
+section CylindersOpen
```
You don't need to have so many sections. The main goal of sections is when you need to change the current context, i.e., the `variable` line.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+
+variable {A G : Type*} [TopologicalSpace A] [DiscreteTopology A]
+
+lemma cylinder_is_closed (U : Finset G) (x : G → A) :
```
```suggestion
lemma isClosed_cylinder (U : Finset G) (x : G → A) :
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+
+/-- Cylinders are open (and, dually, closed) when `A` is discrete. -/
+lemma cylinder_is_open (U : Finset G) (x : G → A) :
+    IsOpen (cylinder (A := A) (G := G) U x) := by
```
```suggestion
    IsOpen (cylinder U x) := by
```
No need to specify A or G here, since they can be deduced from `x`. That's the point of our convention that arguments in definitions should be implicit if they are determined by later explicit arguments.
Same thing below.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+/-- Example: the full shift on alphabet A. -/
+@[to_additive SymbolicDynamics.Ambiant.addFullShift]
+def fullShift (A G) [TopologicalSpace A] [Monoid G] : Subshift A G :=
+{ carrier := Set.univ,
+  isClosed := isClosed_univ,
+  shiftInvariant := by intro _ _ _; simp }
```
indent by two spaces

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,602 @@
+  /-- Finite support of the pattern (subset of `G`). -/
+  support : Finset G
+  /-- The symbol at each point of the support. -/
+  data    : support → A
```
```suggestion
  data : support → A
```
We avoid vertical alignment like that, because it is not robust against refactors, so creates more work for a very small gain.

sgouezel reviewed (commented):
I still have a bunch of comments, of course :-)
A general suggestion: could you try to improve your docstrings to make sure that they are understandable by people who don't already know the notion? (not only the ones I've flagged, but throughout the PR)

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Commit:
refactor(SymbolicDynamics): streamline sections and assumptions

Regroup consecutive sections that had the same variable line (,
, ). Extra assumptions like
are introduced only where first needed. This matches mathlib style and avoids
unnecessary duplication.

Commit:
doc(SymbolicDynamics/FullShift): improve and clarify docstrings

- Add introductory explanations for key notions: configurations, cylinders, patterns, occurrences, forbids, subshifts, languages.
- Make explicit the assumptions needed in docstrings:
  *  for uniqueness in translated patterns (, , etc.).
- Fix minor typos (“ubshift” → “subshift”, “occurence” → “occurrence”).
- Expand several docstrings to highlight intuition and role in symbolic dynamics (domino, forbids, X_F, SFT, languageOn, patternCountOn, etc.).
- Add cross-references between constructions (e.g.  vs ).

Commit:
small docstrings warning and removing A and G from isOpen_cylinder + isClosed_cylinder statement

sgouezel commented:
As you can see, the label has not disappeared. I don't know if -awaiting-author should be on its own line, or in its own message...

Sfgangloff commented:
> As you can see, the label has not disappeared. I don't know if -awaiting-author should be on its own line, or in its own message...

Ah maybe it is because I edited the message to add -awaiting-author

sgouezel reviewed (commented):
Another batch of comments, on the first half of the file. Don't hesitate to also review the second half in light of these comments!

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,716 @@
+variable [TopologicalSpace A]
+
+/-- The right-translation shift is continuous. -/
+@[to_additive] lemma continuous_mulShift (g : G) :
+    Continuous (mulShift (A := A) (G := G) g) := by
+  -- coordinate projections are continuous; composition preserves continuity
+  continuity
```
Got it, thanks

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Commit:
feat(symbolic_dynamics):
- Rename additive image of  to  with
  and update docstrings accordingly; similar modifications for other structures and statements
- Conform to mathlib naming conventions: lowercase structure names;
  rename  →  (and multiplicative variant  if present).

Commit:
style(symbolic_dynamics): switch structure literals to ; minor cleanups

Commit:
docs(symbolic_dynamics): replace inherited docstrings with additive-specific ones

Commit:
style(sd): prefer /

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,746 @@
+  ext h; simp [mulShift]
+
+/-- Composition of right-translation shifts corresponds to multiplication in the group. -/
+@[to_additive shift_mul] lemma mulShift_mul (g₁ g₂ : G) (x : G → A) :
```
why `shift_mul`? I'd rather say `shift_add`. But normally this should be automatic, no need to specify the name by hand.

Commit:
feat: ensure to_additive maps mulShift → shift and MulSubshift → Subshift

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+`h * g` in the original one.
+
+For example, if `G = ℤ` (with addition) and `A = {0,1}`, then
+`mulShift 1 x` is the sequence obtained from `x` by shifting every symbol one
```
```suggestion
For example, if `G = ℤ` (with addition) and `A = {0, 1}`, then
`shift 1 x` is the sequence obtained from `x` by shifting every symbol one
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+The set `U` is called the *support* of the cylinder.
+
+Intuitively, cylinders specify the "letters" on finitely many coordinates, while
+leaving all other coordinates free. For example, in the full shift `{0,1}^ℤ`,
```
```suggestion
leaving all other coordinates free. For example, in the full shift `{0, 1}^ℤ`,
```
Same thing below.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+building blocks: they describe local adjacency conditions between two sites
+of a configuration. -/
+@[to_additive domino]
+def mulDomino {A : Type*} (i j : G) (ai aj : A) : mulPattern A G where
```
Use `Domino` both in the additive and multiplicative situation

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+
+
+attribute [to_additive existing]
+  SymbolicDynamics.FullShift.mulSubshift
```
No need for this, you just need to tag `mulSubshift` with `toAdditive`.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+variable {G : Type*}
+variable [Monoid G]
+
+/-- `p.occursIn x g` means that the finite pattern `p` appears in the configuration `x`
```
Why `occursIn` and not `occursAt`? 'd say `occursIn` means "exists somewhere, but don't specify the position".

Sfgangloff replied:
@sgouezel In the same vein, I think the best option should be occursInAt (occurs in x at g), I thought it would be a bit odd but after all it also makes more sense. I'll change to this, tell me if it sounds ok.

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Commit:
feat(SymbolicDynamics/Basic): refactor additve and multiplicative versions

Commit:
fix arguments unused

Commit:
linters

Commit:
Sync toolchain+lockfile with upstream

Commit:
typo

Commit:
linters

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,731 @@
+In symbolic dynamics, dominos (two-cell patterns) are the simplest nontrivial
+building blocks: they describe local adjacency conditions between two sites
+of a configuration. -/
+def Domino {A : Type*} (i j : G) (ai aj : A) : Pattern A G := by
```
```suggestion
def domino {A : Type*} (i j : G) (ai aj : A) : Pattern A G := by
```
The rule is: if you're defining a type (including Props and structures), then upper case. Otherwise, lower case.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,731 @@
+of a configuration. -/
+def Domino {A : Type*} (i j : G) (ai aj : A) : Pattern A G := by
+  classical
+  refine
```
```suggestion
  exact
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,731 @@
+Formally: if `x` avoids every `p ∈ F` at every position, then for any `h : G`,
+the shifted configuration `shift h x` also avoids every `p ∈ F` at every position. -/]
+lemma mulForbids_shift_invariant (F : Set (Pattern A G)) :
+    ∀ h : G, ∀ x ∈ mulForbids (A := A) (G := G) F, mulShift h x ∈ mulForbids F := by
```
Move `h` left of the colon, and spell this using `MapsTo`.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,731 @@
+This produces a canonical "completion" of the pattern to a configuration,
+filling all unspecified positions with `default`. -/
+@[to_additive patternToOriginConfig]
+def mulPatternToOriginConfig (p : Pattern A G) : G → A :=
```
That's a weird name, because `p` is a `Pattern`, not a `MulPattern` (because there is no `MulPattern`). I'd call this `Pattern.mulExtend` and `Pattern.extend` respectively, to make it possible to use dot notation.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,731 @@
+Uniqueness (and the usual equations such as `patternToConfig p v (w + v) = p.data ⟨w, _⟩`)
+require a right-cancellation hypothesis and are proved in separate lemmas.
+-/]
+noncomputable def mulPatternToConfig (p : Pattern A G) (v : G) : G → A :=
```
Ditto: avoid `mulPattern` and use dot notation.

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Tactic/ToAdditive/GuessName.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Commit:
style: apply review comments (dot notation, MapsTo, variable formatting)

Commit:
refactor: rename mulPatternFromConfig → Pattern.mulFromConfig for consistency

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+so that the theory works not only for groups but also for cancellative monoids.
+
+## Main definitions
+
```
Maybe say that you are only going to describe the additive ones, although multiplicative versions also exist for all the notions (and adapt)? Most of what you write below is additive, right?

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+(or family) of patterns `F` to be *forbidden*. A configuration belongs to the subshift if and only
+it avoids all the forbidden patterns. -/
+@[to_additive forbids
+/-- `forbids F` is the set of configurations that avoid every pattern in `F`.
```
`forbids` of `forbidden`? `F` is forbidden, but I don't really see who or what forbids, so I'd go for the latter.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+
+This is the inverse construction to `Pattern.fromConfig` (which extends a
+finite pattern to a configuration by filling with `default`). -/]
+def mulFromConfig (x : G → A) (U : Finset G) : Pattern A G where
```
ditto, no need to distinguish additive and multiplicative.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+
+This statement uses `[IsRightCancelMul G]` to identify the preimage of `w + v`
+under right-multiplication by `v`. -/]
+lemma mulToConfig_apply_of_mem
```
The name does not match the statement.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+    -- goal: Classical.choose ex = w
+    -- we have h_eq : w' = w and w' := Classical.choose ex
+    simpa [w'] using h_eq     -- <-- use `using h_eq`, only unfold `[w']`
+
```
```suggestion
```
No empty lines in proofs.
Same thing below.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,745 @@
+
+  -- then replace the proof component (same carrier w)
+  have h3 :
+      (⟨w, hw_w⟩ : p.support) = (⟨w, hw⟩ : p.support) := by
```
```suggestion
  have h3 : (⟨w, hw_w⟩ : p.support) = (⟨w, hw⟩ : p.support) := by
```

sgouezel reviewed (commented):
Could you please take my comments into account, but also review the whole PR in light of these comments and what you've learned about mathlib conventions? Thanks!

Commit:
symbolic dynamics: clean up to_additive, namespacing, and docs

Commit:
additional docstring fixes

Sfgangloff commented:
> Could you please take my comments into account, but also review the whole PR in light of these comments and what you've learned about mathlib conventions? Thanks!

Thanks again, I tried to revise more thoroughly this time (and will also next ones). 

-awaiting-author

Commit:
linters

Sfgangloff committed:
Merge remote-tracking branch 'upstream/master' into pr/symbolic-dynamics

Sfgangloff committed:
Update import_all files

Sfgangloff committed:
Renaming additive versions following removal of GuessName.lean

Sfgangloff commented:
> Can you merge master and fix the build? Otherwise, the PR doesn't appear on my queue, so I'm very likely to forget about it. Thanks!

On the way I added some to_additive names because Mathlib/Tactics/ToAdditive/GuessName.lean is removed in the current version of Mathlib.

Sfgangloff commented:
> It doesn't build. The reason is that the syntax for imports has changed: you need to put `module` at the beginning, add `public` to imports, and start your file with `@[expose] public section`. See in another random file how this is done properly.

I was struggling to see why it was not building, thank you for pointing it out!

Sfgangloff committed:
new format for imports

Sfgangloff committed:
Merge remote-tracking branch 'upstream/master' into pr/symbolic-dynamics

Sfgangloff committed:
naming convention corrected for MulFixedSupport

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,749 @@
+  rw [this]
+  refine isClosed_iInter ?_;
+  intro p; refine isClosed_iInter ?_;
+  intro _; refine isClosed_iInter ?_;
```
did you use AI to generate these lines?

Sfgangloff replied:
I rewrote this part to make the structure of the triple intersection more explicit and idiomatic. Thank you for pointing it out.

Sfgangloff committed:
Replace hypothesis [DiscreteTopology A] with [T1Space A] for closed cylinders

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,749 @@
+  simpa [cylinder_eq_set_pi U x] using isOpen_set_pi (U.finite_toSet) (by simp)
+
+/-- Cylinders are closed when `A` is discrete. -/
+lemma isClosed_cylinder (U : Finset G) (x : G → A) :
```
True, but T1Space is needed so that singletons are closed, so I replaced Discrete with T1Space there.

Sfgangloff committed:
T1 Space in comment for cylinder closed lemma

Sfgangloff committed:
adding to_additive abbreviations back

Sfgangloff committed:
removing to_additive names

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,749 @@
+def mulShift (g : G) (x : G → A) : G → A :=
+  fun h => x (h * g)
+
+@[to_additive shift_apply] lemma mulShift_apply (g : G) (x : G → A) (h : G) :
```
Got it, thank you !

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,749 @@
+This shows immediately that `fixedSupport A G U` is finite whenever `U` is. -/
+@[to_additive equivFun
+  /-- Additive version: equivalence between fixed-support additive patterns and functions. -/]
+def mulEquivFun {U : Finset G} :
```
It is true that this is a technical definition, but I am expecting to use it in other files later (like entropy, measures), so I think it is better to define it only once. I gave it a more specific name though (fixedSupport_equiv_fun) to avoid being too generic.

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,749 @@
+  contrapose! hx
+  simpa [mulOccursInAt_mulShift] using hx
+
+/-- The set of patterns with fixed support `U`. -/
```
I have followed your suggestion and replaced the fixed-support machinery by a direct statement that
({p : Pattern A G | p.support = U} : Set (Pattern A G)).Finite
(proved via a local equivalence with U → A). I have also removed the counting definitions (mulLanguageCardOn, mulPatternCountOn) from this PR to keep it focused on the topological and language aspects. The entropy / counting results where these cardinalities really matter will be written in a separate PR.

Sfgangloff committed:
removing counting definitions and use subsets instead of subtypes

Sfgangloff committed:
doc for mulShift

Sfgangloff committed:
Adapted docstrings to last changes

Sfgangloff commented:
> I see that the PR is still marked awaiting-author. Once you're happy with your changes, don't forget to remove the tag, otherwise the PR will not appear on my dashboard. Do you think you're done, or do you still want to do some changes before I have a look?

Thanks for the reminder. It took me a bit of time because I wanted to go through the PR carefully once more and make sure I had corrected everything I could spot - hopefully I didn’t miss anything.
-awaiting-author

sgouezel commented:
I have just noticed that you have removed the awaiting-author tag; since there is a merge conflict, the PR does not appear on my queue. Could you fix the merge conflict?

Sfgangloff committed:
Merge remote-tracking branch 'upstream/master' into pr/symbolic-dynamics

Sfgangloff committed:
linters

Sfgangloff committed:
merged to additive

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,751 @@
+lemma finite_patterns_with_support
+    {A G : Type*} [Fintype A]
+    (U : Finset G) :
+    ({ p : Pattern A G | p.support = U } : Set (Pattern A G)).Finite := by
```
```suggestion
    ({p : Pattern A G | p.support = U}).Finite := by
```
The type ascription doesn't bring anything here, I think.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,751 @@
+  have h_eq :
+      {x | ∀ p ∈ F, ∀ v : G, ¬ p.mulOccursInAt x v}
+        = ⋂ (p : Pattern A G) (hp : p ∈ F) (v : G), {x | ¬ p.mulOccursInAt x v} := by
+    ext x; simp
```
```suggestion
  have h_eq : {x | ∀ p ∈ F, ∀ v : G, ¬ p.mulOccursInAt x v}
    = ⋂ (p : Pattern A G) (hp : p ∈ F) (v : G), {x | ¬ p.mulOccursInAt x v} := by ext; simp
```

sgouezel reviewed (commented):
Sorry for the delay!

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Update Mathlib/Tactic/Translate/ToAdditive.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
Fix underscores in definition and type definition

sgouezel commented on Mathlib/Tactic/Translate/ToAdditive.lean:
```diff
@@ -371,6 +371,11 @@ def abbreviationDict : Std.HashMap String String := .ofList [
   ("divisionAddMonoid", "SubtractionMonoid"),
   ("subNegZeroAddMonoid", "SubNegZeroMonoid"),
   ("modularCharacter", "AddModularCharacter"),
+  ("addShift", "shift"),
```
Have you tried with the left word lower case and the right word upper case, as in the other examples?

Sfgangloff committed:
Fix to additive renaming

Sfgangloff commented:
-awaiting-author

sgouezel commented:
ok. @Sfgangloff, could you summon the bot yourself by copying my line above?

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: botbaki-review[bot] <258399068+botbaki-review[bot]@users.noreply.github.com>

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,743 @@
+of these closed sets over `p ∈ F` and `v ∈ G`. -/]
+lemma isClosed_mulForbidden [DiscreteTopology A] (F : Set (Pattern A G)) :
+    IsClosed (mulForbidden F) := by
+  rw [mulForbidden]
```
Here the comment of the bot is not consistent with proposed modification, which is trivial.

Sfgangloff committed:
suggestions botbaki

Sfgangloff committed:
removed open scoped Classical

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,743 @@
+variable {G : Type*} [Monoid G] [IsRightCancelMul G]
+
+section PatternExtension
+open scoped Classical
```
Ok, thank you ! I removed all open scoped Classical

sgouezel reviewed (commented):
I'm really sorry for the delay, all my apologies. It's a little bit complicated to review a PR like this one, mostly made of definitions, because one can not really see why the definitions are relevant, and if some could be skipped or not -- for instance, you define dominoes but you don't use them yet. I'm sure the applications come later down the road, of course.

I'll try to get the PR inside Mathlib in the near future with your help. But for future development I think it will be more efficient if you build a project downstream of Mathlib, and prove there deep theorems that matter to you. Because proving theorems is what helps get good definitions through refactors imposed by the proofs, and also once you have this you can show off and explain why your work should get into Mathlib.

In any case, I hope that the discussions we had together helped to show you what kind of design choices are important when doing formalization and what style conventions we have, and I hope all these can be put to good use in your project. If you encounter other design questions, don't hesitate to bug me with them on Zulip!

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,743 @@
+We call *configuration* an element of `G → A`.
+
+Given a configuration `x : G → A` and an element `g : G` of the monoid, the shifted configuration
+`mulShift g x` is defined by `(mulShift g x) h = x (h * g)`.
```
You are right, thanks. I was implicitly thinking in the abelian case in which I am usually working on, so that left and right translations coincide. I have corrected it.

Sfgangloff committed:
change from right to left action

Sfgangloff committed:
trigger CI retry for leantar issue

Sfgangloff committed:
Merge remote-tracking branch 'upstream/master' into pr/symbolic-dynamics

Sfgangloff committed:
Removed Fintype A hypothesis from finite_setOf_pattern_support_eq

Sfgangloff committed:
pattern extension at origin

Sfgangloff commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,743 @@
+-- We assume right-cancellation throughout this section for uniqueness of preimages under (_ * v)
+-- or (_ + v).
+
+/-- Turn a finite pattern into a configuration, by extending it with
```
Indeed both extensions can be merged. I changed this.

Thank you for the second suggestion, this does simplify things. I modified the definition of Pattern. However I kept the extension. The reason is a bit subtle and due to the generalization to monoids: at some point (in the statement of mulOccursInAt_eq_cylinder) we need it, otherwise we would need to shift by g^{-1}, which is not necessarily possible. Unless there is a simpler way ?

Sfgangloff committed:
refactor Pattern definition

Sfgangloff committed:
remove section Patternextension

Sfgangloff commented:
-awaiting-author

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,662 @@
+Uniqueness (and the usual equations such as `Pattern.extend p v (v + w) = p.config w`)
+require a left-cancellation hypothesis and are proved in separate lemmas.
+-/]
+noncomputable def Pattern.mulExtend (p : Pattern A G) (v : G) : G → A := by
```
The name seems a little bit strange to me: you're not extending the pattern, you're shifting it. `Pattern.mulShift`?

Sfgangloff replied:
This makes sense, thanks !

Sfgangloff replied:
This seems to create a conflict with the other mulShift definition somehow so I named it Pattern.mulPseudoShift, since it is not a shift in the strictest sense.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,662 @@
+lemma mapsTo_mulShift_mulForbidden {A G : Type*} [Inhabited A] [Monoid G]
+    (F : Set (Pattern A G)) (h : G) :
+    Set.MapsTo (fun x => mulShift h x)
+    (mulForbidden (A := A) (G := G) F) (mulForbidden F) := by
```
```suggestion
    Set.MapsTo (mulShift h) (mulForbidden (A := A) (G := G) F) (mulForbidden F) := by
```

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,662 @@
+
+end Pattern
+
+section OccursInAtEqCylinder
```
No need to open a section, unless there is a good reason for it (typically, variable management)

sgouezel commented:
Only minor comments left, thanks!

Please don't forget to close the discussions in which you considered you've given a satisfactory answer: it makes it possible to focus on the remaining issues.

Sfgangloff committed:
Update Mathlib/Dynamics/SymbolicDynamics/Basic.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

Sfgangloff committed:
mulExtend -> mulShiftExtend + remove section OccursInAtEqCylinder

Sfgangloff committed:
Merge remote-tracking branch 'upstream/master' into pr/symbolic-dynamics

Sfgangloff committed:
rename mulShiftExtend to mulPseudoShift

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,659 @@
+Uniqueness (and the usual equations such as `Pattern.shiftExtend p v (v + w) = p.config w`)
+require a left-cancellation hypothesis and are proved in separate lemmas.
+-/]
+noncomputable def Pattern.mulPseudoShift (p : Pattern A G) (v : G) : G → A := by
```
```suggestion
protected noncomputable def Pattern.mulShift (p : Pattern A G) (v : G) : G → A := by
```
The `pseudo` feels weird, I'd rather go for `mulShift`, but with protected to avoid clashes.

Sfgangloff replied:
Ok ! Changed

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,659 @@
+
+This uses `[IsLeftCancelMul G]` to identify the unique preimage of `v * w`
+under left-multiplication by `v`. -/
+@[to_additive extend_apply_add_left_of_mem
```
Did you forget to update the `extend`, here and elsewhere?

Sfgangloff replied:
Thank you for pointing it out. I thought about it, I guess that happened when reverting the changes multiple times. I checked again, no more extend.

Sfgangloff committed:
pseudoshift -> shift and protected keyword added, extend deleted

Sfgangloff commented:
-awaiting-author

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,658 @@
+  rw [h_eq]
+  -- Now prove that this big intersection is closed.
+  have h_closed :
+      IsClosed (⋂ (p : Pattern A G) (hp : p ∈ F) (v : G), {x | ¬ p.mulOccursInAt x v}) := by
```
This is exactly the main goal. So the simpa line at the end is useless, as is the `have`: just use the current proof to close the main goal.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,658 @@
+
+Formally, `subshift_of_finite_type F` is `Subshift.ofForbidden F` where `F` is a
+`Finset (Pattern A G)`. -/]
+def mulSubshiftOfFiniteType [DiscreteTopology A] (F : Finset (Pattern A G)) : MulSubshift A G :=
```
I would just drop this definition, as using the previous one applied to the coerced finset should always be enough.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,658 @@
+variable {G : Type*}
+
+/-- Patterns with domain exactly `U` form a finite set. -/
+lemma finite_setOf_pattern_support_eq
```
You're using `support` in the lemma name, but `domain` in the statement. I would probably switch the definition of `Pattern` to use `support` as this is a more standard name.

Sfgangloff replied:
Ok, done

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,658 @@
+  -- 1. Upgrade Finite A to Fintype A locally
+  cases nonempty_fintype A
+  -- 2. Explicitly provide the Fintype instance for the Finset U
+  letI : Fintype U := U.fintypeCoeSort
```
```suggestion
```
I don't think this is used.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,658 @@
+    ext p; constructor
+    · intro hp; rcases hp with ⟨p', -, rfl⟩; exact p'.property
+    · intro hp; refine ⟨⟨p, hp⟩, ?_, rfl⟩; simp
+  simpa [himage_eq] using h_image
```
```suggestion
  let : Fintype { p : Pattern A G | p.domain = U } := Fintype.ofEquiv (U → A) e.symm
  apply toFinite
```
You have already done everything, just use `toFinite`.

sgouezel commented on Mathlib/Dynamics/SymbolicDynamics/Basic.lean:
```diff
@@ -0,0 +1,751 @@
+This is the set of all finite patterns obtained by restricting some configuration
+`x ∈ X` to `U`. -/
+@[to_additive languageOn]
+def mulLanguageOn (X : Set (G → A)) (U : Finset G) : Set (Pattern A G) :=
```
Maybe drop the `mul` in the definition, then.

Sfgangloff committed:
mullanguageOn -> languageOn + remove SFT definition + simplification of isClosed_mulForbidden proof

Sfgangloff committed:
simplified proof of finite_setOf_pattern_support_eq

Sfgangloff commented:
-awaiting-author
